{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gbouras13/baktfold/blob/dev/run_baktfold.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QGd2GEI3N-02"
      },
      "source": [
        "#Batkfold\n",
        "\n",
        "[baktfold](https://github.com/gbouras13/baktfold) is a sensitive annotation tool for annotating bacterial genomes, MAGs and plasmids.\n",
        "\n",
        "baktfold uses the [ProstT5](https://github.com/mheinzinger/ProstT5) protein language model to translate hypothetical protein amino acid sequences to the 3Di token alphabet used by [Foldseek](https://github.com/steineggerlab/foldseek). Foldseek is then used to search these against a series of databases of protein structures (AFDB Clusters, SwissProt, CATH and PDB).\n",
        "\n",
        "For now, baktfold takes either the `.json` output file from [batka](https://github.com/oschwengers/bakta) or an input of amino acid sequences (in `.faa` format - these can be any protein sequences you like, not necessarily the output from bakta).\n",
        "\n",
        "If you want to annotate a genome, you will need to run bakta first to get the `.json` file - if you are compute limited, we recommend the awesome [bakta webserver](https://bakta.computational.bio/).\n",
        "\n",
        "* **Before you start, please make sure you the runtime is T4 GPU (or any other kind of GPU if you have $$$), otherwise baktfold won't be installed properly**\n",
        "* To do this, go to the top toolbar, then to Runtime -> Change runtime type -> Hardware accelerator\n",
        "\n",
        "* To run the cells, press the play button on the left side\n",
        "* Cells 1, 2 and 3 installs baktfold and downloads the databases/models.\n",
        "* Once cells 1-3 have been run, you can re-run Cell 4 (to run `baktfold run` for genomes) and/or Cell 5 (to run `baktfold proteins` for proteins) to annotate different inputs as many times as you would like\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ii39RG8eOZUx"
      },
      "outputs": [],
      "source": [
        "\n",
        "#@title 1. Install miniforge\n",
        "\n",
        "#@markdown This cell installs miniforge\n",
        "\n",
        "%%bash\n",
        "\n",
        "set -e\n",
        "\n",
        "PYTHON_VERSION=$(python3 -c \"import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')\")\n",
        "\n",
        "echo \"python version ${PYTHON_VERSION}\"\n",
        "\n",
        "if [ ! -f CONDA_READY ]; then\n",
        "  echo \"installing miniforge\"\n",
        "\n",
        "  # miniforge 25.9.1 introduces some issue - latest as of 7 Nov 2025 - see https://github.com/gbouras13/phold/issues/106\n",
        "  # issue is fixed if you use the previous release (25.3.1)\n",
        "  #wget -qnc https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh\n",
        "  wget -qnc https://github.com/conda-forge/miniforge/releases/download/25.3.1-0/Miniforge3-25.3.1-0-Linux-x86_64.sh\n",
        "  bash Miniforge3-25.3.1-0-Linux-x86_64.sh -bfp /usr/local 2>&1 1>/dev/null\n",
        "  conda config --set auto_update_conda false\n",
        "  touch CONDA_READY\n",
        "fi\n",
        "\n",
        "pip install --upgrade matplotlib matplotlib-inline\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tfltpbZ_QLfZ"
      },
      "outputs": [],
      "source": [
        "#@title 2. Install baktfold\n",
        "\n",
        "#@markdown This cell installs baktfold. It will take a few minutes. Please be patient\n",
        "\n",
        "BAKTFOLD_VERSION=\"0.0.2\"\n",
        "\n",
        "# add paths\n",
        "import sys\n",
        "sys.path.append(\"/usr/local/bin\")\n",
        "\n",
        "# Update environment variables for shell usage\n",
        "import os\n",
        "os.environ[\"PATH\"] = \"/usr/local/bin:\" + os.environ[\"PATH\"]\n",
        "\n",
        "# create envs\n",
        "# pharokka isn't compatible with Python 3.13 (Google Colab default)\n",
        "# so it needs a separate env\n",
        "from pathlib import Path\n",
        "flag_file = Path(\"BAKTFOLD_READY\")\n",
        "if not flag_file.exists():\n",
        "  !conda install -y -c bioconda baktfold=={BAKTFOLD_VERSION} pytorch=*=cuda*\n",
        "  # Touch the flag file\n",
        "  flag_file.touch()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Sjdpu6R-Kig9"
      },
      "outputs": [],
      "source": [
        "#@title 3. Download baktfold databases\n",
        "\n",
        "#@markdown This cell downloads the baktfold database. It will take some time (5-15 minutes probably depending on Zenodo's traffic). Please be patient. Perhaps go for a walk or have a coffee or tea.\n",
        "\n",
        "\n",
        "%%time\n",
        "\n",
        "print(\"Downloading baktfold database. This will take a few minutes. Please be patient :)\")\n",
        "!baktfold install -d baktfold_db -t 8 --foldseek-gpu\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "9QfjP3q-Q04f",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5d90f1d3-0bfd-4a70-d8a8-76a57fabcf49"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Input file assembly.json exists\n",
            "Running baktfold\n",
            "baktfold completed successfully.\n",
            "Your output is in output_baktfold.\n",
            "Zipping the output directory so you can download it all in one go.\n",
            "Output directory has been zipped to output_baktfold.zip\n",
            "CPU times: user 90.4 ms, sys: 4.93 ms, total: 95.3 ms\n",
            "Wall time: 1min 42s\n"
          ]
        }
      ],
      "source": [
        "#@title 4. baktfold run (bakta json input)\n",
        "\n",
        "#@markdown First, upload your bakta json file.\n",
        "\n",
        "#@markdown Click on the folder icon to the left and use the file upload button.\n",
        "\n",
        "#@markdown Once it is uploaded, write the file name in the INPUT_FILE field on the right.\n",
        "\n",
        "#@markdown Then provide a directory for baktfold's output using BAKTFOLD_OUT_DIR.\n",
        "#@markdown The default is 'output_baktfold'.\n",
        "\n",
        "#@markdown You can also provide a prefix for your output files with BAKTFOLD_PREFIX.\n",
        "#@markdown If you provide nothing it will default to 'baktfold'.\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory.\n",
        "#@markdown This may be useful if your earlier pharokka run has crashed for whatever reason.\n",
        "\n",
        "#@markdown The results of Baktfold will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is BAKTFOLD_OUT_DIR.zip, where BAKTFOLD_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "import sys\n",
        "import subprocess\n",
        "import zipfile\n",
        "INPUT_FILE = '' #@param {type:\"string\"}\n",
        "\n",
        "if os.path.exists(INPUT_FILE):\n",
        "    print(f\"Input file {INPUT_FILE} exists\")\n",
        "else:\n",
        "    print(f\"Error: File {INPUT_FILE} does not exist\")\n",
        "    print(f\"Please check the spelling and that you have uploaded it correctly\")\n",
        "    sys.exit(1)\n",
        "\n",
        "BAKTFOLD_OUT_DIR = 'output_baktfold'  #@param {type:\"string\"}\n",
        "BAKTFOLD_PREFIX = 'baktfold'  #@param {type:\"string\"}\n",
        "FORCE = True  #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "# Construct the command\n",
        "command = (\n",
        "    f'baktfold run -d baktfold_db -i {INPUT_FILE} -t 4 '\n",
        "    f'-o {BAKTFOLD_OUT_DIR} -p {BAKTFOLD_PREFIX} --foldseek-gpu'\n",
        ")\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running baktfold\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"baktfold completed successfully.\")\n",
        "    print(f\"Your output is in {BAKTFOLD_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{BAKTFOLD_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(BAKTFOLD_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), BAKTFOLD_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 4. baktfold proteins (protein input)\n",
        "\n",
        "#@markdown First, upload your file containing protein amino acid sequences (`.faa` format).\n",
        "\n",
        "#@markdown Click on the folder icon to the left and use the file upload button.\n",
        "\n",
        "#@markdown Once it is uploaded, write the file name in the INPUT_FILE field on the right.\n",
        "\n",
        "#@markdown Then provide a directory for baktfold's output using BAKTFOLD_PROTEINS_OUT_DIR.\n",
        "#@markdown The default is 'output_baktfold_proteins'.\n",
        "\n",
        "#@markdown You can also provide a prefix for your output files with BAKTFOLD_PREFIX.\n",
        "#@markdown If you provide nothing it will default to 'baktfold'.\n",
        "\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory.\n",
        "#@markdown This may be useful if your earlier pharokka run has crashed for whatever reason.\n",
        "\n",
        "#@markdown The results of Baktfold will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is BAKTFOLD_PROTEINS_OUT_DIR.zip, where BAKTFOLD_PROTEINS_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "import sys\n",
        "import subprocess\n",
        "import zipfile\n",
        "INPUT_FILE = '' #@param {type:\"string\"}\n",
        "\n",
        "if os.path.exists(INPUT_FILE):\n",
        "    print(f\"Input file {INPUT_FILE} exists\")\n",
        "else:\n",
        "    print(f\"Error: File {INPUT_FILE} does not exist\")\n",
        "    print(f\"Please check the spelling and that you have uploaded it correctly\")\n",
        "    sys.exit(1)\n",
        "\n",
        "BAKTFOLD_PROTEINS_OUT_DIR = 'output_baktfold_proteins'  #@param {type:\"string\"}\n",
        "BAKTFOLD_PREFIX = 'baktfold'  #@param {type:\"string\"}\n",
        "FORCE = True  #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "# Construct the command\n",
        "command = (\n",
        "    f'baktfold proteins -d baktfold_db -i {INPUT_FILE} -t 4 '\n",
        "    f'-o {BAKTFOLD_PROTEINS_OUT_DIR} -p {BAKTFOLD_PREFIX} --foldseek-gpu'\n",
        ")\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running baktfold\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"baktfold completed successfully.\")\n",
        "    print(f\"Your output is in {BAKTFOLD_PROTEINS_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{BAKTFOLD_PROTEINS_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(BAKTFOLD_PROTEINS_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), BAKTFOLD_PROTEINS_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "1eUYA415PNRX",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d2be93f0-12f5-4352-d87d-7e64646bcbe1"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Input file assembly.hypotheticals.faa exists\n",
            "Running baktfold\n",
            "baktfold completed successfully.\n",
            "Your output is in output_baktfold_proteins.\n",
            "Zipping the output directory so you can download it all in one go.\n",
            "Output directory has been zipped to output_baktfold_proteins.zip\n",
            "CPU times: user 8.73 ms, sys: 8.25 ms, total: 17 ms\n",
            "Wall time: 1min 52s\n"
          ]
        }
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "authorship_tag": "ABX9TyN9fm6I5AWClvGUc55SBQdl",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "nbformat": 4,
  "nbformat_minor": 0
}